File: scripts/generate-renders.py
#!/usr/bin/env python3
"""
CleanWalker Robotics — AI Render Generator v2.0
Uses the Replicate API to generate photorealistic robot renders.
Model: Seedream 4.5 (bytedance/seedream-4.5) at 16:9
All prompts start with the CANONICAL robot description from docs/design/robot-design-spec.md
"""

import json
import os
import sys
import time
import urllib.request
import urllib.error

API_TOKEN = os.environ.get("REPLICATE_API_TOKEN")
if not API_TOKEN:
    print("ERROR: REPLICATE_API_TOKEN environment variable not set")
    sys.exit(1)
MODEL = "bytedance/seedream-4.5"
API_BASE = "https://api.replicate.com/v1"
OUTPUT_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                          "apps", "web", "public", "renders")

# Canonical robot description — MUST start every prompt
CORE = (
    "A quadrupedal robot dog (similar to Unitree Go2) with a matte dark olive-green body, "
    "bright green LED strips on each side, four articulated legs with rubber foot pads, "
    "a single articulated robotic arm with mechanical gripper mounted on the front top, "
    "and an X-shaped scissor frame on its back holding open a semi-transparent black trash bag "
    "with a square rim at the top. The robot is about medium-dog sized, ~60cm at the body. "
    "Industrial-modern aesthetic."
)

# All 14 render scenes — each gets CORE prepended
RENDERS = [
    {
        "filename": "hero-park.png",
        "scene": (
            " The robot is in a clean urban park, its mechanical gripper arm reaching down and "
            "picking up a plastic bottle from the grass. The semi-transparent trash bag on its "
            "back is partially filled with collected litter visible through the material. "
            "Morning golden hour lighting, shallow depth of field, green grass and trees in "
            "background. Photorealistic product photography, 8K, ultra detailed."
        ),
    },
    {
        "filename": "hero-sidewalk.png",
        "scene": (
            " The robot is walking on a European city sidewalk with cobblestone street. Four "
            "legs in dynamic walking pose. Modern architecture in the background. Overcast sky, "
            "soft diffused lighting. Professional product photography, studio quality, 8K."
        ),
    },
    {
        "filename": "hero-fleet.png",
        "scene": (
            " Wide-angle photo showing three of these identical robots working together in a "
            "large urban park, picking up litter. One robot is picking up a crushed can with "
            "its mechanical gripper arm, another is walking with its bag nearly full of litter, "
            "and the third has its scissor frame lowered flat, dropping a sealed full bag at "
            "the curb. Drone perspective, morning light, professional marketing photo. "
            "Ultra detailed, 8K."
        ),
    },
    {
        "filename": "detail-gripper.png",
        "scene": (
            " Extreme close-up focused on the robot's mechanical gripper with 2-3 metal "
            "fingers firmly grasping a crushed aluminum can. The gripper is attached to the "
            "single articulated robotic arm. In the background, slightly out of focus, the "
            "raised X-shaped scissor bag frame holds the open trash bag with collected litter "
            "visible inside. Shallow depth of field, studio lighting, product photography "
            "style, 8K macro photography."
        ),
    },
    {
        "filename": "detail-sensors.png",
        "scene": (
            " Front view close-up of the robot's face showing two square bright green LED "
            "eyes (camera/sensor windows) on the front panel. The matte dark olive-green body "
            "panels have clean industrial design with slightly rounded chamfered edges. Behind "
            "and above, the raised X-shaped scissor bag frame is partially visible holding "
            "the open trash bag. Studio product photography with soft rim lighting, 8K."
        ),
    },
    {
        "filename": "detail-side-profile.png",
        "scene": (
            " Clean side-profile product photo on a white/light gray studio background. "
            "Standing pose, four articulated legs visible with all joints and servo motors, "
            "the single robotic arm with mechanical gripper visible at front top, and the "
            "X-shaped scissor bag frame raised on its back holding the open semi-transparent "
            "trash bag — the robot's most distinctive feature. Professional product "
            "photography, even lighting, no harsh shadows, 8K."
        ),
    },
    {
        "filename": "detail-charging-dock.png",
        "scene": (
            " The robot is parked on a small charging dock platform in an urban park setting. "
            "The dock is a simple weatherproof platform with a small rain canopy. Bright green "
            "LED strips glowing as a charging indicator. The X-shaped scissor bag frame is "
            "lowered flat against the body in compact transport mode, giving the robot a low "
            "sleek profile. Evening golden hour lighting, 8K."
        ),
    },
    {
        "filename": "lifestyle-city-worker.png",
        "scene": (
            " A city parks maintenance worker in a high-visibility vest stands next to the "
            "robot, checking a tablet showing fleet status. The robot is about knee-height, "
            "with its X-shaped scissor bag frame raised and holding an open semi-transparent "
            "trash bag half-full of collected litter. They are in a well-maintained urban park. "
            "The worker looks pleased. Natural daylight, editorial photography style, 8K."
        ),
    },
    {
        "filename": "lifestyle-night-ops.png",
        "scene": (
            " The robot is operating on a city street at night. The bright green LED strips "
            "illuminate the sidewalk around it. Two square green LED eyes on the front glow "
            "visibly. Street lamps in the background. The robot is mid-stride with its raised "
            "X-shaped scissor bag frame holding the open trash bag. The mechanical gripper arm "
            "is pressing a piece of litter down into the bag. Cinematic lighting, shallow "
            "depth of field, moody atmosphere, 8K."
        ),
    },
    {
        "filename": "lifestyle-before-after.png",
        "scene": (
            " Split image composition: Left side shows a park area with scattered litter "
            "(plastic bottles, cans, food wrappers) on the ground. Right side shows the same "
            "park area pristine and clean, with the robot visible in the distance, its "
            "X-shaped scissor bag frame lowered flat, a sealed full bag dropped neatly at the "
            "curb behind it. Before/after comparison, bright daylight, professional editorial "
            "photography, 8K."
        ),
    },
    {
        "filename": "tech-exploded-view.png",
        "scene": (
            " Technical exploded view diagram on a white background. The robot's components "
            "are floating in space showing: dark olive-green weatherproof enclosure panels, "
            "aluminum frame chassis, 12 servo actuator modules for the four legs, stereo "
            "camera system, LiDAR sensor puck, compute module, 48V battery pack, single "
            "mechanical gripper arm assembly, X-shaped scissor bag frame with square rim, and "
            "bright green LED strip modules. Clean technical illustration style with thin "
            "leader lines and component labels. Professional, minimalist, 8K."
        ),
    },
    {
        "filename": "tech-dashboard-mockup.png",
        "scene": (
            " Split composition: Left side shows the robot in a park with its raised X-shaped "
            "scissor bag frame holding an open trash bag, its mechanical gripper arm actively "
            "picking up a piece of litter. Right side shows a fleet management dashboard on a "
            "large monitor displaying a map with robot positions, bags collected statistics, "
            "and battery levels. Modern office environment. Professional product marketing "
            "photo, 8K."
        ),
    },
    {
        "filename": "component-actuator.png",
        "scene": (
            " Close-up detail shot focusing on one of the robot's compact brushless DC servo "
            "actuator modules, removed from a leg joint and placed on a reflective dark "
            "surface. Cylindrical black anodized aluminum housing, about 70mm diameter, with "
            "CAN bus connector cable. The full robot is visible in soft focus in the "
            "background, showing its olive-green body and X-shaped scissor bag frame. Subtle "
            "rim lighting, clean industrial premium feel, studio product photography, 8K."
        ),
    },
    {
        "filename": "component-pcb.png",
        "scene": (
            " Close-up detail shot focusing on the robot's custom green PCB (printed circuit "
            "board) with SMD components, CAN bus connectors, and power distribution rails. "
            "Clean solder joints, professional assembly. On a dark matte surface with dramatic "
            "side lighting. The full robot is visible in soft focus in the background showing "
            "its dark olive-green body and raised X-shaped scissor bag frame. 8K macro "
            "photography."
        ),
    },
]


def api_request(method, url, data=None):
    """Make an authenticated API request to Replicate."""
    headers = {
        "Authorization": f"Bearer {API_TOKEN}",
        "Content-Type": "application/json",
    }
    body = json.dumps(data).encode() if data else None
    req = urllib.request.Request(url, data=body, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            return json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        error_body = e.read().decode() if e.fp else "no body"
        print(f"  HTTP {e.code}: {error_body[:200]}")
        return None
    except Exception as e:
        print(f"  Request error: {e}")
        return None


def create_prediction(prompt):
    """Create a prediction on Replicate."""
    url = f"{API_BASE}/models/{MODEL}/predictions"
    data = {
        "input": {
            "prompt": prompt,
            "aspect_ratio": "16:9",
        }
    }
    return api_request("POST", url, data)


def poll_prediction(prediction_id, max_wait=300):
    """Poll a prediction until it completes or fails."""
    url = f"{API_BASE}/predictions/{prediction_id}"
    start = time.time()
    while time.time() - start < max_wait:
        result = api_request("GET", url)
        if result is None:
            time.sleep(5)
            continue
        status = result.get("status")
        if status == "succeeded":
            return result
        elif status in ("failed", "canceled"):
            print(f"  Prediction {status}: {result.get('error', 'unknown')}")
            return None
        time.sleep(4)
    print(f"  Prediction timed out after {max_wait}s")
    return None


def download_file(url, filepath):
    """Download a file from URL to filepath. Returns file size in bytes."""
    req = urllib.request.Request(url)
    with urllib.request.urlopen(req, timeout=120) as resp:
        data = resp.read()
    with open(filepath, "wb") as f:
        f.write(data)
    return len(data)


def generate_render(render, attempt=1):
    """Generate a single render. Returns (success, filesize)."""
    filename = render["filename"]
    prompt = CORE + render["scene"]
    filepath = os.path.join(OUTPUT_DIR, filename)

    # For retries, add emphasis on key design elements
    if attempt > 1:
        prompt += (
            " IMPORTANT: The robot body color is dark olive-green (not black, not gray). "
            "Bright green LED strips must be visible on the sides. It has exactly one "
            "mechanical metal gripper arm (not silicone, not soft). The X-shaped scissor "
            "frame with semi-transparent black trash bag on the back must be clearly visible."
        )

    print(f"\n{'='*60}")
    print(f"[{filename}] Attempt {attempt}")
    sys.stdout.flush()

    # Create prediction
    result = create_prediction(prompt)
    if result is None:
        return False, 0

    pred_id = result.get("id")
    if not pred_id:
        print(f"  No prediction ID returned")
        return False, 0
    print(f"  Prediction: {pred_id}")
    sys.stdout.flush()

    # Poll for completion
    result = poll_prediction(pred_id)
    if result is None:
        return False, 0

    # Get output URL
    output = result.get("output")
    if isinstance(output, list):
        output_url = output[0] if output else None
    elif isinstance(output, str):
        output_url = output
    else:
        print(f"  Unexpected output format: {type(output)}")
        return False, 0

    if not output_url:
        print(f"  No output URL")
        return False, 0

    # Download
    try:
        filesize = download_file(output_url, filepath)
        size_kb = filesize / 1024
        print(f"  Downloaded: {size_kb:.0f} KB -> {filepath}")
        sys.stdout.flush()
    except Exception as e:
        print(f"  Download error: {e}")
        return False, 0

    # Validate
    if filesize < 200_000:
        print(f"  WARNING: Suspiciously small ({size_kb:.0f} KB < 200 KB)")
        return False, filesize

    print(f"  OK ({size_kb:.0f} KB)")
    return True, filesize


def main():
    print("=" * 60)
    print("CleanWalker Render Generation v2.0")
    print(f"Model: Seedream 4.5 via Replicate")
    print(f"Renders: {len(RENDERS)}")
    print(f"Output: {OUTPUT_DIR}")
    print("=" * 60)
    sys.stdout.flush()

    os.makedirs(OUTPUT_DIR, exist_ok=True)

    stats = {"first_try": 0, "retry": 0, "failed": 0}
    details = []

    for i, render in enumerate(RENDERS):
        print(f"\n>>> [{i+1}/{len(RENDERS)}] {render['filename']}")
        sys.stdout.flush()

        success, filesize = generate_render(render, attempt=1)

        if not success:
            print(f"  RETRYING {render['filename']}...")
            sys.stdout.flush()
            success, filesize = generate_render(render, attempt=2)
            if success:
                stats["retry"] += 1
            else:
                stats["failed"] += 1
        else:
            stats["first_try"] += 1

        details.append({
            "filename": render["filename"],
            "success": success,
            "filesize": filesize,
        })

    # Summary
    print("\n" + "=" * 60)
    print("GENERATION SUMMARY")
    print("=" * 60)
    print(f"First-try success:  {stats['first_try']}")
    print(f"Retry success:      {stats['retry']}")
    print(f"Failed:             {stats['failed']}")
    print(f"Total:              {len(RENDERS)}")
    print()
    for d in details:
        status = "OK" if d["success"] else "FAIL"
        kb = d["filesize"] / 1024 if d["filesize"] else 0
        print(f"  [{status:4s}] {d['filename']:35s} {kb:6.0f} KB")
    print("=" * 60)
    sys.stdout.flush()

    return stats


if __name__ == "__main__":
    stats = main()
    sys.exit(0 if stats["failed"] == 0 else 1)

File: scripts/generate-renders.mjs
#!/usr/bin/env node
/**
 * Generate all 14 CleanWalker robot renders via Replicate API (seedream-4.5)
 * Usage: node scripts/generate-renders.mjs [--only hero-park,hero-sidewalk]
 */

import { writeFile } from 'fs/promises';
import { join } from 'path';

const REPLICATE_API_TOKEN = process.env.REPLICATE_API_TOKEN;
if (!REPLICATE_API_TOKEN) {
  console.error('ERROR: REPLICATE_API_TOKEN not set');
  process.exit(1);
}

const OUTPUT_DIR = join(import.meta.dirname, '..', 'apps', 'web', 'public', 'renders');

const CORE = `A custom-designed quadrupedal robot with a smooth matte dark olive-green body with softly rounded panel edges. The body is a flat horizontal rectangular enclosure, roughly 45cm long, 30cm wide, and 15cm tall. The head is a compact sensor housing module at the front with a flat face featuring two square bright green LED eye panels — not an organic dog head but a sleek sensor module with gentle contouring. Bright green LED light strips on the outward-facing sides of the upper leg segments only. Four articulated legs roughly the same length as the body, mammalian stance, with rubber foot pads. A tall multi-joint robotic arm with mechanical gripper rises from the front top of the body, extending about 40cm above the body with visible shoulder, elbow, and wrist joints. On the center of the flat back, a black cylindrical bag roll dispenser mounted horizontally with its axis running left-to-right across the body width. At the rear edge of the body, a rectangular dark metal tube frame is hinged open at 135 degrees from the back surface, extending backward and upward — it is the same width as the body and about half the body length deep. A heavy-duty black trash bag hangs freely from this frame, draping down behind the robot — no basket, no cage, no mesh. Smooth clean modern surfaces throughout. No text, no logos, no branding, no watermarks.`;

const CORE_CHARGING = `A custom-designed quadrupedal robot with a smooth matte dark olive-green body with softly rounded panel edges. The body is a flat horizontal rectangular enclosure, roughly 45cm long, 30cm wide, and 15cm tall. The head is a compact sensor housing module at the front with a flat face featuring two square bright green LED eye panels — not an organic dog head but a sleek sensor module with gentle contouring. Bright green LED light strips on the outward-facing sides of the upper leg segments only. Four articulated legs roughly the same length as the body, mammalian stance, with rubber foot pads. A tall multi-joint robotic arm with mechanical gripper rises from the front top of the body, extending about 40cm above the body with visible shoulder, elbow, and wrist joints. On the center of the flat back, a black cylindrical bag roll dispenser mounted horizontally with its axis running left-to-right across the body width. Smooth clean modern surfaces throughout. No text, no logos, no branding, no watermarks.`;

const CORE_EXPLODED = `A custom-designed quadrupedal robot with a smooth matte dark olive-green body with softly rounded panel edges. The body is a flat horizontal rectangular enclosure, roughly 45cm long, 30cm wide, and 15cm tall. The head is a compact sensor housing module at the front with a flat face featuring two square bright green LED eye panels. Smooth clean modern surfaces throughout. No text, no logos, no branding, no watermarks.`;

const CORE_COMPONENT = `A custom-designed quadrupedal robot with a smooth matte dark olive-green body with softly rounded panel edges. The body is a flat horizontal rectangular enclosure, roughly 45cm long, 30cm wide, and 15cm tall. The head is a compact sensor housing module at the front with a flat face featuring two square bright green LED eye panels. Bright green LED light strips on the outward-facing sides of the upper leg segments only. Four articulated legs roughly the same length as the body, mammalian stance. A tall multi-joint robotic arm with mechanical gripper rises from the front top. At the rear, a rectangular dark metal tube frame is hinged open with a heavy-duty black trash bag hanging freely. Smooth clean modern surfaces throughout. No text, no logos, no branding, no watermarks.`;

const RENDERS = [
  {
    name: 'hero-park',
    prompt: `${CORE} The robot is in a clean urban park, its mechanical gripper arm reaching down and picking up a plastic bottle from the grass. The heavy-duty black trash bag hanging from the rear folding frame is partially filled with collected litter. Morning golden hour lighting, shallow depth of field, green grass and trees in background. Photorealistic product photography, 8K, ultra detailed.`
  },
  {
    name: 'hero-sidewalk',
    prompt: `${CORE} The robot is walking on a European city sidewalk with cobblestone street. Four legs in dynamic walking pose. Modern architecture in the background. Overcast sky, soft diffused lighting. Professional product photography, studio quality, 8K.`
  },
  {
    name: 'hero-fleet',
    prompt: `${CORE} Wide-angle photo showing three of these identical robots working together in a large urban park, picking up litter. One robot is picking up a crushed can with its mechanical gripper arm, another is walking with its bag nearly full of litter, and the third has its folding frame folded down flat against the body, a sealed full bag dropped neatly at the curb behind it. Drone perspective, morning light, professional marketing photo. Ultra detailed, 8K.`
  },
  {
    name: 'detail-gripper',
    prompt: `${CORE} Extreme close-up focused on the robot's mechanical gripper with 2-3 metal fingers firmly grasping a crushed aluminum can. The gripper is attached to the single articulated robotic arm. In the background, slightly out of focus, the rear folding frame holds the open trash bag with collected litter visible inside. Shallow depth of field, studio lighting, product photography style, 8K macro photography.`
  },
  {
    name: 'detail-sensors',
    prompt: `${CORE} Front view close-up of the robot's sensor housing head module showing two square bright green LED eye panels on the flat front face. The matte dark olive-green body panels have clean industrial design with softly rounded chamfered edges. Behind and above, the rear folding frame is partially visible holding the open trash bag. Studio product photography with soft rim lighting, 8K.`
  },
  {
    name: 'detail-side-profile',
    prompt: `${CORE} Clean side-profile product photo on a white/light gray studio background. Standing pose, four articulated legs visible with all joints and actuator housings, the single robotic arm with mechanical gripper visible at front top, and the rectangular folding frame hinged open at 135 degrees from the rear of the body with the heavy-duty black trash bag hanging freely — the robot's most distinctive feature. On the back, the horizontal bag roll dispenser is visible between the arm mount and the frame hinge. Professional product photography, even lighting, no harsh shadows, 8K.`
  },
  {
    name: 'detail-charging-dock',
    prompt: `${CORE_CHARGING} The robot is parked on a small charging dock platform in an urban park setting. The dock is a simple weatherproof platform with a small rain canopy. Bright green LED strips glowing on the upper legs as a charging indicator. The rectangular folding frame at the rear is folded down flat against the body in compact transport mode, giving the robot a low sleek profile. Evening golden hour lighting, 8K.`
  },
  {
    name: 'lifestyle-city-worker',
    prompt: `${CORE} A city parks maintenance worker in a high-visibility vest stands next to the robot, checking a tablet showing fleet status. The robot is about knee-height, with its folding frame open and the heavy-duty black trash bag half-full of collected litter hanging behind it. They are in a well-maintained urban park. The worker looks pleased. Natural daylight, editorial photography style, 8K.`
  },
  {
    name: 'lifestyle-night-ops',
    prompt: `${CORE} The robot is operating on a city street at night. The bright green LED strips on the upper legs illuminate the sidewalk around it. Two square green LED eye panels on the sensor housing head glow visibly. Street lamps in the background. The robot is mid-stride with its folding frame holding the open trash bag behind the body. The mechanical gripper arm is pressing a piece of litter down into the bag. Cinematic lighting, shallow depth of field, moody atmosphere, 8K.`
  },
  {
    name: 'lifestyle-before-after',
    prompt: `${CORE.replace('A heavy-duty black trash bag hangs freely from this frame, draping down behind the robot — no basket, no cage, no mesh. ', '')} Split image composition: Left side shows a park area with scattered litter (plastic bottles, cans, food wrappers) on the ground. Right side shows the same park area pristine and clean, with the robot visible in the distance, its folding frame folded down flat against the body, a sealed full bag dropped neatly at the curb behind it. Before/after comparison, bright daylight, professional editorial photography, 8K.`
  },
  {
    name: 'tech-exploded-view',
    prompt: `${CORE_EXPLODED} Technical exploded view diagram on a white background. The robot's components are floating in space showing: dark olive-green weatherproof enclosure panels with rounded edges, aluminum frame chassis, 12 servo actuator modules for the four legs, stereo camera system in the sensor housing head, LiDAR sensor puck, compute module, 48V battery pack, single mechanical gripper arm assembly, horizontal black cylindrical bag roll dispenser, rectangular dark metal folding bag frame with hinge mechanism, bright green LED strip modules for the upper legs, and rubber foot pads. Clean technical illustration style with thin leader lines. Professional, minimalist, 8K.`
  },
  {
    name: 'tech-dashboard-mockup',
    prompt: `${CORE} Split composition: Left side shows the robot in a park with its folding frame holding an open trash bag, its mechanical gripper arm actively picking up a piece of litter. Right side shows a fleet management dashboard on a large monitor displaying a map with robot positions, bags collected statistics, and battery levels. Modern office environment. Professional product marketing photo, 8K.`
  },
  {
    name: 'component-actuator',
    prompt: `${CORE_COMPONENT} Close-up detail shot focusing on one of the robot's compact brushless DC servo actuator modules, removed from a leg joint and placed on a reflective dark surface. Cylindrical black anodized aluminum housing, about 70mm diameter, with CAN bus connector cable. The full robot is visible in soft focus in the background, showing its olive-green body and the folding frame with trash bag behind. Subtle rim lighting, clean industrial premium feel, studio product photography, 8K.`
  },
  {
    name: 'component-pcb',
    prompt: `${CORE_COMPONENT} Close-up detail shot focusing on the robot's custom green PCB (printed circuit board) with SMD components, CAN bus connectors, and power distribution rails. Clean solder joints, professional assembly. On a dark matte surface with dramatic side lighting. The full robot is visible in soft focus in the background showing its dark olive-green body and the folding frame with trash bag. 8K macro photography.`
  }
];

async function createPrediction(prompt) {
  const res = await fetch('https://api.replicate.com/v1/models/bytedance/seedream-4.5/predictions', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${REPLICATE_API_TOKEN}`,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      input: {
        prompt,
        aspect_ratio: '16:9',
        num_outputs: 1
      }
    })
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Create prediction failed (${res.status}): ${text}`);
  }
  return res.json();
}

async function pollPrediction(id) {
  const url = `https://api.replicate.com/v1/predictions/${id}`;
  while (true) {
    const res = await fetch(url, {
      headers: { 'Authorization': `Bearer ${REPLICATE_API_TOKEN}` }
    });
    const data = await res.json();
    if (data.status === 'succeeded') return data;
    if (data.status === 'failed' || data.status === 'canceled') {
      throw new Error(`Prediction ${id} ${data.status}: ${data.error || 'unknown error'}`);
    }
    // Wait 3 seconds before polling again
    await new Promise(r => setTimeout(r, 3000));
  }
}

async function downloadImage(url, filePath) {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Download failed: ${res.status}`);
  const buffer = Buffer.from(await res.arrayBuffer());
  await writeFile(filePath, buffer);
  return buffer.length;
}

async function generateOne(render) {
  const outPath = join(OUTPUT_DIR, `${render.name}.png`);
  console.log(`[${render.name}] Creating prediction...`);
  const prediction = await createPrediction(render.prompt);
  console.log(`[${render.name}] Prediction ${prediction.id} created, polling...`);
  const result = await pollPrediction(prediction.id);
  const imageUrl = Array.isArray(result.output) ? result.output[0] : result.output;
  console.log(`[${render.name}] Downloading image...`);
  const size = await downloadImage(imageUrl, outPath);
  console.log(`[${render.name}] Saved ${outPath} (${(size / 1024).toFixed(0)} KB)`);
  return { name: render.name, path: outPath, size };
}

// Parse --only flag
const onlyArg = process.argv.find(a => a.startsWith('--only='));
let filter = null;
if (onlyArg) {
  filter = onlyArg.split('=')[1].split(',');
  console.log(`Generating only: ${filter.join(', ')}`);
}

const toGenerate = filter ? RENDERS.filter(r => filter.includes(r.name)) : RENDERS;

// Generate in batches of 4 to avoid rate limiting
const BATCH_SIZE = 4;
const results = [];

console.log(`\nGenerating ${toGenerate.length} renders via Replicate API (seedream-4.5)...\n`);

for (let i = 0; i < toGenerate.length; i += BATCH_SIZE) {
  const batch = toGenerate.slice(i, i + BATCH_SIZE);
  console.log(`--- Batch ${Math.floor(i / BATCH_SIZE) + 1} (${batch.map(r => r.name).join(', ')}) ---`);
  const batchResults = await Promise.all(batch.map(r => generateOne(r)));
  results.push(...batchResults);
}

console.log(`\n=== DONE: ${results.length} renders generated ===`);
results.forEach(r => console.log(`  ${r.name}: ${(r.size / 1024).toFixed(0)} KB`));
